modules/apps/portal/portal-messaging/src/main/java/com/liferay/portal/messaging/internal/DefaultMessageBus.java

```diff
@@ -300,6 +301,12 @@ public void sendMessage(String destinationName, Message message) {
 
 		message.setDestinationName(destinationName);
 
+		if (destination instanceof BaseAsyncDestination) {
+			message.put(
```

Comment:
@tinatian we should do the threadlocals capturing and _asyncDestinationInterceptor.intercept() right here, to avoid setting the interceptor into the message. And the _asyncDestinationInterceptor should always intercept the message even the BaseAsyncDestination currently has no listeners, the current solution skips the interception when there is no listener which is not right.

Reply:
@shuyangzhou 
the threadlocals capturing is only needed by asynchronous destination, if we put it here, the threadlocal capturing and replaying will be separated, currently all threadlocal logic is in same class, putting it in different class is hard to maintain, and also comparing setting one property to message, this doesn’t make more sense to me, cause the threadlocal logic should be always consistent, it’s better we put it in one class.
 for the listener issue, if you see it from destination side, it is true that we should always intercept no matter there’s listener or not, cause it is named as a destination intercepter, but if from the function of destination, that is to invoke listener, if there’s no listener, the destination is actually not used at all, it is same as no destination, then intercepting only when listener existed also makes sense. also considering our real case in change tracking, what we want to intercept is not destination, it is actually listeners. if you really care about the listener issue, we can also delay the listener checking into destination to avoid it. 

Tina